rfcs/1440_generic_planner_node.md

```diff
@@ -0,0 +1,237 @@
+# RFC#1440: Configuration-based Generic Planner Node
+
+## Objective
+
+Make the node graph easier to set up, reason about and amenable to static
+analysis.
+
+## Background
+
+An Oak application creates an initial node and calls the entry point on that
+node. A typical pattern is to use this initial node as a "Planner Node",
+responsible for creating all other nodes and channels. The creation of the nodes
+an channels is done through code. For simple cases it is easy to figure out what
+the graph can look like by inspecting the code, but it becomes increasingly
+difficult as the complexity of the logic of the planner node increases to cope
+with more complex node graphs. Another drawback of the code-based dynamic node
+creation is that we have to run the code to use the introspection mechanism to
+visualise the node graph.
+
+Generating the node graph automatically based on a configuration/manifest file
+should simplify this. It should be easier to figure out what the node graph will
```

Comment:
As a historical aside, we had exactly this prior to #340.

Reply:
The plan was to start with the simple case of static-only nodes and channels and do a follow-up for dynamic label-based node-creation. My thoughts so far on the dynamic label-based creation have been very much along these lines.

I also planned on doing this in two steps. The first step would be to allow generic node and channel templates which would extract some part of the label as a parameter, which would then be used to make concrete instances of the node definitions from the templates and create the nodes and channels. This should support cases like the trusted database that requires a single node per client label.

The next step would be to support defining templates for complete sub-graphs.

Seeing that dynamic node creation seems to be one of the main points people are interested in I will add a detailed design to this RFC.

Reply:
I have added a detailed design for the template-based dynamic sub-graph creation and removed it from the future work section. PTAL.